File: 2022-2023-2/Compilation-Principle-and-Design/Lab-1/Python/src/main.py
import time
import random

from algorithm import quicksort as qs

def sort(arr, length):
    
    sortArr = qs.quicksort(arr, 0, length - 1)
    
    return sortArr

if __name__ == '__main__':
    
    avg_time = 0
    
    for i in range(100):
        arr = list(range(1, 100001))
        random.shuffle(arr)
    
        length = len(arr)
        
        start_time = time.time()
        sortArr    = sort(arr, length)
        end_time   = time.time()
        
        total_time = end_time - start_time
        
        avg_time = (avg_time * i + total_time) / (i + 1)
        
        print("{:>3d} : Run Time: {:.2f} ms".format(i + 1, total_time * 1000))
        
    print("Avg Time: {:.4f} ms".format(avg_time * 1000))

File: 2022-2023-2/Compilation-Principle-and-Design/Lab-1/Java/src/Main.java
package src;

import java.util.Random;
import java.util.Scanner;

import src.algorithm.quicksort;

public class Main {
	
	public static int[] randomArr(int length) {
		
		int[] arr = new int[100001];
		
		for (int i = 0; i < length; i++) {
			
			arr[i] = i + 1;
		}
		
		Random random = new Random();
		for (int j = length - 1; j > 0; j--) {
			
			int index = random.nextInt(length);
			
			int temp   = arr[index];
			arr[index] = arr[j]    ;
			arr[j]     = temp      ;
		}
		
		return arr;
	}
	
	public static void main(String [] args) {
		
		long start_time, end_time;
		long total_time, avg_time;
		
		avg_time = 0;
		
		quicksort qs = new quicksort();
		
		for (int i = 1; i <= 100; i++) {
			
			int   len = 100000;
			int[] arr = randomArr(len);
			
			start_time = System.currentTimeMillis();
			arr        = qs.quicksort(arr, 0, len - 1);
			end_time   = System.currentTimeMillis();
		
			total_time = end_time - start_time;
			avg_time   = avg_time + total_time;
			
			System.out.printf("%3d : Run Time: %2d ms\n", i, total_time);
		}
		
		System.out.printf("Avg Time: %d ms", avg_time / 100);
		
		Scanner input = new Scanner(System.in);
		System.out.print("\nPress Enter to quit...");
		input.nextLine();
		System.exit(0);
	}
}